crates/runtime/src/status.rs

```diff
@@ -43,49 +47,126 @@ impl Display for ComponentStatus {
+        self.update_component_status(format!("model:{model_name}"), status);
+        metrics::models::STATUS.record(
+            status as u64,
+            &[Key::from_static_str("model").string(model_name)],
+        );
+    }
+
+    pub fn update_tool(&self, tool_name: &str, status: ComponentStatus) {
+        let tool_name = tool_name.to_string();
+        self.update_component_status(format!("tool:{tool_name}"), status);
+        metrics::tools::STATUS.record(
+            status as u64,
+            &[Key::from_static_str("tool").string(tool_name)],
+        );
+    }
 
-pub fn update_embedding(model_name: &str, status: ComponentStatus) {
-    let model_name = model_name.to_string();
-    metrics::embeddings::STATUS.record(
-        status as u64,
-        &[Key::from_static_str("model").string(model_name)],
-    );
+    pub fn update_llm(&self, model_name: &str, status: ComponentStatus) {
+        let model_name = model_name.to_string();
+        self.update_component_status(format!("llm:{model_name}"), status);
+        metrics::llms::STATUS.record(
+            status as u64,
+            &[Key::from_static_str("model").string(model_name)],
+        );
+    }
+
+    pub fn update_embedding(&self, model_name: &str, status: ComponentStatus) {
+        let model_name = model_name.to_string();
+        self.update_component_status(format!("embedding:{model_name}"), status);
+        metrics::embeddings::STATUS.record(
+            status as u64,
+            &[Key::from_static_str("model").string(model_name)],
+        );
+    }
+
+    /// Checks if all registered components have been ready at least once.
+    ///
+    /// This function returns `true` if all components that have ever been registered
+    /// have reached the `Ready` state at least once.
+    /// Once this state is reached, it will continue to return `true` regardless of the
+    /// current state of any component.
+    ///
+    /// This is intentionally conservative - in the accelerated datasets case, we can
+    /// continue to serve data from the acceleration layer even if the source dataset
+    /// is in an error state.
+    ///
+    /// Returns `false` if:
+    /// - No components have been registered yet.
+    /// - There are one or more registered components that have never been in the `Ready` state.
+    #[must_use]
+    pub fn is_ready(&self) -> bool {
+        let statuses = match self.statuses.read() {
+            Ok(guard) => guard,
+            Err(poisoned) => poisoned.into_inner(),
+        };
+        let ever_ready = match self.ever_ready_components.read() {
+            Ok(guard) => guard,
+            Err(poisoned) => poisoned.into_inner(),
+        };
+
+        if statuses.is_empty() {
+            return false; // No components registered yet
+        }
+
+        // Check if all registered components have been ready at least once
+        statuses
+            .keys()
+            .all(|component| ever_ready.contains(component))
+    }
```

Comment:
This is the logic to check if the runtime is ready

Reply:
Learning purpose: what will be behavior for dynamically added components - will it be added to components and the status will be changed to not ready until new dataset/component is loaded?

Reply:
That shouldn't affect the `/ready` API response, as adding components at runtime is mainly a local convenience. And even if we did want to support loading components dynamically as a production scenario, loading a new component shouldn't cause the API to break for existing components that have already initialized. The `/ready` API is mainly for making sure that on initial startup we indicate when all of the components defined in the spicepod are ready to serve requests, and its safe to put the pod into the load balancer